samcli/lib/samlib/resource_metadata_normalizer.py

```diff
@@ -144,3 +154,65 @@ def _update_resource_metadata(metadata, updated_values):
         """
         for key, val in updated_values.items():
             metadata[key] = val
+
+    @staticmethod
+    def get_resource_id(resource_properties, logical_id):
+        """
+        Get unique id for a resource.
+        for any resource, the resource id can be the customer defined id if exist, if not exist it can be the
+        cdk-defined resource id, or the logical id if the resource id is not found.
+
+        Parameters
+        ----------
+        resource_properties dict
+            Properties of this resource
+        logical_id str
+            LogicalID of the resource
+
+        Returns
+        -------
+        str
+            The unique function id
+        """
+        resource_metadata = resource_properties.get("Metadata", {})
+        customer_defined_id = resource_metadata.get(SAM_RESOURCE_ID_KEY)
+
+        if isinstance(customer_defined_id, str) and customer_defined_id:
+            LOG.debug(
+                "Sam customer defined id is more priority than other IDs. Customer defined id for resource %s is %s",
+                logical_id,
+                customer_defined_id,
+            )
+            return customer_defined_id
+
+        resource_cdk_path = resource_metadata.get(RESOURCE_CDK_PATH_METADATA_KEY)
```

Comment:
This method and filename is named generic (`resource_metadata_normalizer` and `get_resource_id`), but method itself contains CDK specific stuff. This feels like it should be a generic (let's say `ResourceMetadataNormalizer`) class and we should be assigning different types depending on the project type (`CfnResourceMetadataNormalizer` or `CdkResourceMetadataNormalizer`), what do you think?

Reply:
I am working on another PR that is related to this part (IaC ids) to use it in other commands. Can I work on this refactoring after I merge this PR, and the one I am working on, as I do not want to miss things, it will be easier to refactor everything together